### PR TITLE
fixing maximization condition for queens generator

### PR DESCRIPTION
**Pull request by athomas125**
Was using the queens generator with a runner (see here for example https://github.com/hiive/mlrose/blob/master/problem_examples.ipynb)

It is built to work as a minimization problem. However it seems like everything has to be phrased as a maximization problem for this assignment. In order to do this, I set:

problem = QueensGenerator().generate(seed=seed, size=8)

problem.maximize = 1.0

Here however I ran into an issue where the examples will all run for the max number of examples, because the stopping condition in queens_opt (which is what is given from the QueensGenerator()) is simply int(self.get_fitness()) == 0

obviously this doesn't work for a maximization problem, so I updated it. The condition when you maximize the n queens is n choose 2, which is n!/(2!)(n-2!), which simplifies to n * (n-1) / 2

### CODE REVIEW

**Comment by AlexWendland** [comment]
I agree with the change, but I think for the problem to make sense as a maximising problem you will need to add to fitness/queens.py a dependency on the maximising parameter. Something like 
```
if self.maximize:
    fitness = (self.length * (self.length-1)) / 2 - fitness
return fitness
```

File: mlrose_hiive/opt_probs/queens_opt.py [hunk 1]
```diff
@@ -36,4 +36,6 @@ def __init__(self, length=None, fitness_fn=None, maximize=False,
         self.set_state(state)
 
     def can_stop(self):
+        if self.maximize == 1.0:
```
Comment (stevenjackson121) [inline mlrose_hiive/opt_probs/queens_opt.py:39]: ```if self.maximize:```
is typically preferable to 

```if self.maximize == 1.0``` 

for clarity and brevity's sake. The argument's default value is a boolean, and realistically we only care about its truth value.

Comment (athomas125) [reply mlrose_hiive/opt_probs/queens_opt.py:39]: "The argument's default value is a boolean" - this is not true
look at line 47 in _OptProb initializer. 
maximize is passed in as a boolean, but self.maximize is not a boolean, I copied the code snippet below

        if maximize:
            self.maximize = 1.0
        else:
            self.maximize = -1.0

This means that if self.maximize will always evaluate to true, hence the necessity of self.maximize == 1.0

Comment (hiive) [reply mlrose_hiive/opt_probs/queens_opt.py:39]: Is this right that the problem *has* to be stated as a maximization problem?

If so, then it's changed from when I took the course. If not, then I'm loathe to make a code change that isn't actually necessary.

**Comment by AlexWendland** [comment]
I have also been thinking about this change, and a better upper bound on the score is the following. 
If length is 1, 0, 
if length 2, 1, 
for length 3 or higher it is 3 * (length-2)
This can be proved by restating the fitness function in terms of matching queens to the right of any queen (as matches on the left of one queen is on the right of the matching queen). For each queen you can match on any of the 2 diagonals or row, so at most 3 matches. The reason we have length-2, for the most right queen it can't match, for the second most right queen it can match at most 1 queen, for the 3rd most right queen it can match at most 2 queens. Putting this all together we get 3*(length-3) + 2 + 1 = 3*(length - 2).  
I have made a PR with this change and also allowing for the maximise parameter to be passed down into k-queens: https://github.com/hiive/mlrose/pull/30 (it also includes an update to k-colour problem in a similar vein)
The PR is not that neat as it is just the merge from my fork for this assignment.

File: mlrose_hiive/opt_probs/queens_opt.py (hunk 1 above)
Comment (AlexWendland) [reply mlrose_hiive/opt_probs/queens_opt.py:39]: Yes, for the assignment this year all problems have to be stated as maximisation problems.

**Comment by hiive** [comment]
Already included in other pull request.

**Pull request closed without merging**